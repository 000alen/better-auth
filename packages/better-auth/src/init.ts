import { createKyselyAdapter } from "./adapters/kysely";
import { getAdapter } from "./adapters/utils";
import { createInternalAdapter } from "./db";
import type { BetterAuthOptions } from "./types";
import { getBaseURL } from "./utils/base-url";
import {
	type BetterAuthCookies,
	createCookieGetter,
	getCookies,
} from "./utils/cookies";
import { createLogger } from "./utils/logger";

export const init = (options: BetterAuthOptions) => {
	const adapter = getAdapter(options);
	const db = createKyselyAdapter(options);
	const baseURL = getBaseURL(options.baseURL);

	return {
		options: {
			...options,
			baseURL,
		},
		baseURL: baseURL || "",
		session: {
			updateAge: options.session?.updateAge || 24 * 60 * 60, // 24 hours
			expiresIn: options.session?.expiresIn || 60 * 60 * 24 * 7, // 7 days
		},
		secret:
			options.secret ||
			process.env.BETTER_AUTH_SECRET ||
			process.env.AUTH_SECRET ||
			"better-auth-secret-123456789",
		authCookies: getCookies(options),
		logger: createLogger({
			disabled: options.disableLog,
		}),
		db,
		adapter: adapter,
		internalAdapter: createInternalAdapter(adapter, options),
		createAuthCookie: createCookieGetter(options),
	};
};

export type AuthContext = {
	options: BetterAuthOptions;
	baseURL: string;
	authCookies: BetterAuthCookies;
	logger: ReturnType<typeof createLogger>;
	db: ReturnType<typeof createKyselyAdapter>;
	adapter: ReturnType<typeof getAdapter>;
	internalAdapter: ReturnType<typeof createInternalAdapter>;
	createAuthCookie: ReturnType<typeof createCookieGetter>;
	secret: string;
	session: {
		updateAge: number;
		expiresIn: number;
	};
};
