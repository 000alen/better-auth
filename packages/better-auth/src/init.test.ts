import { describe, expect, it } from "vitest";
import { init } from "./init";

describe("init", async () => {
	it("should match config", () => {
		const res = init({
			database: {
				provider: "sqlite",
				url: ":memory:",
			},
		});
		expect(res).toMatchSnapshot();
	});

	it("should mount internal plugins", () => {
		const res = init({
			database: {
				provider: "sqlite",
				url: ":memory:",
			},
			advanced: {
				crossSubDomainCookies: {
					enabled: true,
				},
			},
		});
		expect(res.options.plugins).toHaveLength(1);
		expect(res.options.plugins[0]["id"]).toBe("cross-subdomain-cookies");
	});

	it("should execute plugins init", () => {
		let changedCtx = {
			baseURL: "http://test.test",
		};
		const res = init({
			database: {
				provider: "sqlite",
				url: ":memory:",
			},
			plugins: [
				{
					id: "test",
					init: (options) => {
						return changedCtx;
					},
				},
			],
		});
		expect(res).toMatchObject(changedCtx);
	});
});
