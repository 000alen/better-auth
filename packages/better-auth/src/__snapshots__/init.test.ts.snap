// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`init > should match config 1`] = `Promise {}`;
